lib/sms/sms_status_report.c

```diff
@@ -0,0 +1,386 @@
+ */
+#define TP_SRQ_SMS_SUBMIT           0x00
+#define TP_SRQ_SMS_COMMAND          0x01
+
+/**
+ * @brief First byte of SMS-STATUS-REPORT PDU in 3GPP TS 23.040 chapter 9.2.2.3
+ */
+struct pdu_status_report_header {
+	uint8_t mti : 2;  /** @brief TP-Message-Type-Indicator */
+	uint8_t mms : 1;  /** @brief TP-More-Messages-to-Send */
+	uint8_t lp : 1;   /** @brief TP-Loop-Prevention */
+	uint8_t:1;        /** @brief Empty bit */
+	uint8_t srq : 1;  /** @brief TP-Status-Report-Qualifier */
+	uint8_t udhi : 1; /** @brief TP-User-Data-Header-Indicator */
+};
+
+/**
+ * @brief TP-Status in 3GPP TS 23.040 chapter 9.2.3.15
+ */
+struct pdu_status_report_status {
+	uint8_t status : 7;  /** @brief TP-Status */
+	uint8_t:1;           /** @brief Empty bit */
+};
+
+/**
+ * @brief SMS-STATUS-REPORT PDU fields specified in 3GPP TS 23.040 chapter 9.2.2.1
+ */
+struct pdu_status_report_data {
+	struct pdu_status_report_header header; /** @brief First byte of header */
+	uint8_t                         mr;     /** @brief TP-Message-Reference */
+	struct sms_address              ra;     /** @brief TP-Recipient-Address */
+	struct sms_time                 scts;   /** @brief TP-Service-Centre-Time-Stamp */
+	struct sms_time                 dt;     /** @brief TP-Discharge-Time */
+	struct pdu_status_report_status st;     /** @brief TP-STATUS */
+};
+
+/** extern functions */
+uint8_t swap_nibbles(uint8_t value);
+uint8_t semioctet_to_dec(uint8_t value);
+void convert_number_to_str(uint8_t *number, uint8_t number_length, char *str_number);
+
+/**
+ * @brief Decode first byte of SMS-STATUS-REPORT header as specified in 3GPP TS 23.040 9.2.2.3.
+ *
+ * @param[in,out] parser Parser instance.
+ * @param[in] buf Buffer containing PDU and pointing to this field.
+ *
+ * @return Number of parsed bytes.
+ */
+static int parse_pdu_status_report_header(struct parser *parser, uint8_t *buf)
+{
+	struct pdu_status_report_data * const pdata = parser->data;
+
+	pdata->header = *((struct pdu_status_report_header *)buf);
+
+	LOG_DBG("SMS header 1st byte: 0x%02X", *buf);
+
+	LOG_DBG("TP-Message-Type-Indicator: %d", pdata->header.mti);
+	LOG_DBG("TP-More-Messages-to-Send: %d", pdata->header.mms);
+	LOG_DBG("TP-Loop-Prevention: %d", pdata->header.lp);
+	LOG_DBG("TP-Status-Report-Qualifier: %d", pdata->header.srq);
+	LOG_DBG("TP-User-Data-Header-Indicator: %d", pdata->header.udhi);
+
+	/* Only parse SMS-STATUS-REPORT for SMS-SUBMIT */
+	if (pdata->header.mti != TP_MTI_SMS_STATUS_REPORT ||
+	    pdata->header.srq != TP_SRQ_SMS_SUBMIT) {
+		return -EINVAL;
+	}
+
+	return 1;
+}
+
+/**
+ * @brief Decode TP-Message-Reference as specified in 3GPP TS 23.040 Section 9.2.3.6.
+ *
+ * @param[in,out] parser Parser instance.
+ * @param[in] buf Buffer containing PDU and pointing to this field.
+ *
+ * @return Number of parsed bytes.
+ */
+static int parse_pdu_mr_field(struct parser *parser, uint8_t *buf)
+{
+	struct pdu_status_report_data * const pdata = parser->data;
+
+	pdata->mr = (uint8_t)*buf;
```

Comment:
I just want to point out that as this is not in `struct sms_status_report_header`, and we are not actually setting TP_MR in SMS-SUBMIT (except for concatenated message), you don't know which message this status report applies to. It may not be a big concern but just want to make sure you are aware of it.

Reply:
Closing this comment as SMS library changes have been removed from this PR.